datahub-web-react/src/app/lineage/utils/constructTree.ts

```diff
@@ -1,16 +1,87 @@
+import { Entity } from '../../../types.generated';
 import EntityRegistry from '../../entity/EntityRegistry';
-import { Direction, EntityAndType, FetchedEntities, NodeData } from '../types';
+import { Direction, EntityAndType, FetchedEntities, FetchedEntity, NodeData, UpdatedLineages } from '../types';
 import constructFetchedNode, { shouldIncludeChildEntity } from './constructFetchedNode';
+import extendAsyncEntities from './extendAsyncEntities';
+
+function createEntityAndType(entity: Entity) {
+    return {
+        type: entity.type,
+        entity,
+    } as EntityAndType;
+}
+
+function updateFetchedEntity(fetchedEntity: FetchedEntity, updatedLineages: UpdatedLineages) {
+    if (!(fetchedEntity.urn in updatedLineages)) {
+        return fetchedEntity;
+    }
+    const updatedLineage = updatedLineages[fetchedEntity.urn];
+    let updatedEntity = fetchedEntity;
+    const entitiesToAdd = updatedLineage.entitiesToAdd.map((entity) => createEntityAndType(entity));
+    if (updatedLineage.lineageDirection === Direction.Upstream) {
+        updatedEntity = {
+            ...updatedEntity,
+            upstreamChildren: [
+                ...(updatedEntity.upstreamChildren || []).filter(
+                    (child) => !updatedLineage.urnsToRemove.includes(child.entity.urn),
+                ),
+                ...entitiesToAdd.filter(
+                    (entity) =>
+                        !(updatedEntity.upstreamChildren || []).find((child) => child.entity.urn === entity.entity.urn),
+                ),
+            ],
+        };
+    } else {
+        updatedEntity = {
+            ...updatedEntity,
+            downstreamChildren: [
+                ...(updatedEntity.downstreamChildren || []).filter(
+                    (child) => !updatedLineage.urnsToRemove.includes(child.entity.urn),
+                ),
+                ...entitiesToAdd.filter(
+                    (entity) =>
+                        !(updatedEntity.downstreamChildren || []).find(
+                            (child) => child.entity.urn === entity.entity.urn,
+                        ),
+                ),
+            ],
+        };
+    }
+
+    return updatedEntity;
+}
 
 export default function constructTree(
     entityAndType: EntityAndType | null | undefined,
     fetchedEntities: FetchedEntities,
     direction: Direction,
     entityRegistry: EntityRegistry,
+    updatedLineages: UpdatedLineages,
 ): NodeData {
     if (!entityAndType?.entity) return { name: 'loading...', children: [] };
     const constructedNodes = {};
 
+    let updatedFetchedEntities = fetchedEntities;
+    Object.entries(updatedFetchedEntities).forEach((entry) => {
```

Comment:
this block is difficult to follow.. as is as tjhe code in the file, unfortunately.

Reply:
yeah :( i was really struggling in here - as mentioned our lineage code can use a face lift